arangod/Pregel/Worker/Worker.cpp

```diff
@@ -332,16 +352,37 @@ void Worker<V, E, M>::_startProcessing() {
   Scheduler* scheduler = SchedulerFeature::SCHEDULER;
 
   _feature.metrics()->pregelNumberOfThreads->fetch_add(1);
+  _initializeMessageCaches();
 
   auto self = shared_from_this();
   scheduler->queue(RequestLane::INTERNAL_LOW, [self, this] {
     if (_state != WorkerState::COMPUTING) {
       LOG_PREGEL("f0e3d", WARN) << "Execution aborted prematurely.";
       return;
     }
-    if (_processVertices() && _state == WorkerState::COMPUTING) {
-      _finishedProcessing();  // last thread turns the lights out
+
+    std::vector<futures::Future<Result>> futures;
+    // first build up all future requests
+    for (auto [idx, quiver] : enumerate(_quivers)) {
+      TRI_ASSERT(_state == WorkerState::COMPUTING);
+      futures.emplace_back(_processVertices(idx, quiver));
     }
+
+    if (!futures.empty()) {
+      // wait for all futures to finalize
+      auto futureResponses = futures::collectAll(futures).get();
+      for (auto const& fResponse : futureResponses) {
+        auto& result = fResponse.get();
+        if (result.fail()) {
+          LOG_PREGEL("ee2ac", WARN) << "Execution aborted prematurely.";
```

Comment:
It would be quite nice to have some kind of indication of what error happened if we have a way to get it.

Reply:
also added "result.errorMessage()" to be bassed into the "LOG_PREGEL " statement